### ISSUE TITLE
HDR support

### ISSUE THREAD

**Issue by thomas7428** [Blinue/Magpie#55]
Currently MagPie can not be used with HDR enabled content. 
I suppose that this is due to a lack of support from WinRT and GDI so it would be awesome to add the support of HDR to one of those.
Cyberpunk 2077 with HDR (FSR and WinRT):
![image](https://user-images.githubusercontent.com/44552992/128880023-3d82d042-1b29-49d1-8951-3fbd7a1c7ba5.png)

Cyberpunk 2077 without HDR (FSR and WinRT): 
![image](https://user-images.githubusercontent.com/44552992/128880093-66c38629-64bd-42bb-b685-3c5d16c5eabd.png)

Vermintide 2 with auto HDR (windows 11 feature) (FSR and WinRT):
![image](https://user-images.githubusercontent.com/44552992/128880619-31269fe7-accf-47ff-8d8d-ec61f3ec7be4.png)

Vermintide 2 without HDR (FSR and WinRT):
![image](https://user-images.githubusercontent.com/44552992/128881067-5dac18bb-75b5-485c-b92a-bc7e8bdc62fc.png)

**Comment by Blinue** [comment]
Magpie does not support HDR, and this feature may not be added for a long time.

**Comment by aaronth07** [comment]
As an alternative, Lossless Scaling on Steam has worked for me pretty well. Hope i'm not breaking any rules recommending an alternative considering the dev has said that support will not be added, and so the program is completely unusable for many people and games.

**Comment by Blinue** [comment]
LS is a great program, I would recommend it if Magpie's FSR doesn't meet your needs. 

Contributions are very welcome as my monitor doesn't support HDR.

**Comment by bloc97** [comment]
I also don't have a HDR monitor, but I think it might be cool to add inverse-tonemapping methods if HDR was to be supported. This way you could play SDR games on HDR screens using Magpie. For instance, it is possible to train high quality inverse tonemapping neural networks.

**Comment by Blinue** [comment]
I'm not familiar with HDR, but it would be really cool if Magpie could add HDR support for games that don't support it natively.

**Comment by bloc97** [comment]
Apparently windows has Auto-HDR which automatically transforms SDR/LDR content into HDR content. I can't say how good it is though.

**Comment by mirh** [comment]
Autohdr hooks into the rendering pipeline to convert it (SpecialK is also another tool that can do the same) the games becoming legit native HDR then. 

But even to do tonemapping yourself you'd have to somewhat tinker with the application process.. because in order not to have windows apply its own dumb one, I suppose the only way is to report to the system that the window is HDR-aware/ready.

**Comment by aaronth07** [comment]
No it doesn't. Auto HDR just takes the SDR output and uses an AI conversion
to tonemap to HDR. It doesn't hook into the game any more than it needs to
(which is, it needs to be DX11+ for the game window to output HDR).
Technically Auto HDR could be activated system wide, but I guess Microsoft
doesn't do that because they fear it won't look good in all games.

**Comment by mirh** [comment]
Damn, I always [forget that](https://twitter.com/blueisviolet/status/1384918002265624577), sorry. 
Anyhow, SpecialK still does what I said.

**Comment by Cutano** [comment]
FYI, I found a useful article about Windows HDR standard and how to capture HDR content.
[high-dynamic-range/capturing-hdr-and-wcg-screen-content](https://learn.microsoft.com/en-us/windows/win32/direct3darticles/high-dynamic-range#capturing-hdr-and-wcg-screen-content)

**Comment by CharonM72** [comment]
Update to this: it _is_ possible to get Magpie to output in HDR (not natively per se, but using Windows Auto HDR, which is almost as good). Unfortunately unless Linux or Mac have something similar, it won't work on those platforms.

TL;DR:
- Make sure Auto HDR is turned on in Windows settings in System -> Display -> HDR
- Download the latest version of this: https://github.com/7gxycn08/ForceAutoHDR/releases
- Run it and click "Add Game EXE"
- Navigate to where you have Magpie installed and add Magpie.exe. You should get confirmation that it was added successfully. Do NOT add the game that you wish to use it on
- Restart Magpie (if necessary)
- Anything you upscale using Magpie should now turn to HDR _once it is upscaled_ (it will still be SDR before upscaling)

**Why this works**
Magpie does not have the ability to natively upscale HDR content. This is likely due to the fact that the upscale models themselves cannot handle HDR color spaces (would be my guess). Magpie also does not (currently) have the ability to tonemap (intelligently convert HDR color space to SDR color space). Therefore, if Magpie is fed HDR content, it treats it like SDR content, which means the much larger color space of HDR will be crushed down to SDR levels, causing extreme washing out of color. This also applies to games or content where HDR is added using Windows Auto HDR, because Magpie pulls the image _after_ Windows applies HDR, so the same issue occurs.

However, the way Magpie works is that when you upscale with Magpie, you are not actually viewing the original game, you are viewing Magpie's output. Therefore, the solution is to apply HDR _after_ Magpie does its upscaling, onto Magpie's own output (that's why you tell Windows to add HDR to Magpie and not the game itself). By using the ForceAutoHDR tool above, you can force Windows to treat Magpie as Auto HDR-capable, and Windows will intelligently use its proprietary machine-learning based algorithm to automatically tonemap SDR content to HDR color space for a supported HDR-capable screen.

Anyway, hope that helps some people out there.

**Comment by aaronth07** [comment]
> Damn, I always [forget that](https://twitter.com/blueisviolet/status/1384918002265624577), sorry. Anyhow, SpecialK still does what I said.

Not to bump this again, but I might add that SpecialK always attempts to injects into the game and therefore doesn't work with any game with an anticheat. Unlike Magpie which i'm assuming only transforms the Window or uses dxgi duplication or similar. 

You are correct that SpecialK is an extremely useful tool, if it had a version with only window controls it would be even more useful. But regardless of HDR "upscaling" or anything similar, just very basic passthrough support would be greatly appreciated as now the program is practically unusable with a large portion of content (since even enabling HDR in Windows will overblow the highlights in MagPie unless you set the brightness nigh-unusably low). 

I would just recommend remembering to replace the shaders with HDR enabled/linear color space versions, as hilariously and very disappointingly, despite having HDR passthrough support most shader in Lossless scaling do not actually support HDR, leading to artifacts that make those shaders unusable. Luckily, most "name brand" shaders (e.g. FSR, and NIS) should already have versions that support HDR output, if it's not even a basic toggle in the script.

**Comment by mirh** [comment]
Tbf just so many even native games have dogshit HDR, and you have to mod them with RenoDX anyway (also there's RTX HDR now?)
Anyway yes, I guess magpie+[reshadeAHR](https://github.com/EndlesslyFlowering/AutoHDR-ReShade) could be the only option for the most cursed competitive multiplayer games.

Now that even `Windows.Graphics.Capture` is supported (DXGI duplication should be good too since 1.6?) it shouldn't be *that hard* to add.

**Comment by aaronth07** [comment]
So is this even on the radar? I might add that another unique ability of Magpie, over Shaderglass, is the integration with the native Window. Shaderglass requires either duplication or glass mode, neither of which work well with resizing, mouse input passthrough, etc. Whereas lossless scaling doesn't work well (or at all) with windowed applications, nor does it have the shader variety. The ability to natively upscale any application with not only resolution but also dynamic range would be insanely useful. Imagine being able to use a browser or photo viewer with FSR plus autoHDR as an example. Magpie's superior and "native" feeling integration for upscaling modes, window selection, etc., in my opinion, make it a far more attractive choice than Shaderglass. Shaderglass does due things that Magpie does not (e.g. the "glass" mode), but that's fine since they fulfill different roles and niches. In another thread you (or the dev) said that Magpie's primary purpose is not for high performance game upscaling, that's understandable and even preferably, what drew me to Magpie is that it wasn't focused on gaming alone. However, by not making HDR a priority, you are excluding 50% of the market, or more. Most new monitors today support some form of HDR, and HDR adjustment/settings/filters is why I install Reshade or launch Shaderglass 99% of the time. The other 1% being niche use cases like upscaling with Shaderglass, or sharpening Shaders.

I see that this was moved to "in progress" (I seem to come back here every couple of months when researching the same/a similar topic), I'm not sure if you finally obtained an HDR monitor or the status of HDR integration, but I would be happy to help with testing if necessary.
